src/CBT/Environment.hs

```diff
@@ -1,70 +1,76 @@
-{-# LANGUAGE MonoLocalBinds #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# OPTIONS -Wno-orphans #-}
 
 module CBT.Environment
-  ( Environment(..)
+  ( AppT
+  , Environment(..)
   , HasEnvironment
-  , IOEnvironment
+  , WithEnv
   , getEnvironment
   , newDefaultEnvironment
   , onDebug
   , runDefaultEnvironment
+  , runDefaultEnvironmentLog
   , runEnvironment
+  , runEnvironmentLog
   )
 where
 
 import CBT.IncrementalState
 import CBT.Prelude
 import CBT.Types
+import Control.Monad.Catch (MonadCatch, MonadThrow)
 import Control.Monad.IO.Unlift (MonadUnliftIO(..), wrappedWithRunInIO)
 import Control.Monad.Reader (MonadReader, ReaderT, ask, runReaderT)
-import Control.Monad.Trans.Class (MonadTrans(lift))
+import Control.Monad.Trans.Class (MonadTrans, lift)
 import Data.Maybe (isJust)
 
 import qualified CBT.IncrementalState as IncrementalState
 import qualified Colog
 import qualified System.Environment   as Environment
 
-type IOEnvironment = Environment (App IO)
+type WithEnv m env = (Colog.WithLog env Colog.Message m, HasEnvironment m, MonadUnliftIO m)
 
-data Environment m = Environment
-  { debug     :: Bool
-  , builds    :: IncrementalState ImageName
-  , logAction :: Colog.LogAction m Colog.Message
+data Environment = Environment
+  { debug  :: Bool
+  , builds :: IncrementalState ImageName ImageBuildError ()
   }
 
-instance MonadIO m => Colog.HasLog (Environment m) Colog.Message m where
-  getLogAction                        = logAction
-  setLogAction logAction' environment = environment { logAction = logAction' }
-
-newtype App m a = App { runApp :: ReaderT (Environment (App m)) m a }
+newtype AppT m a = AppT { runAppT :: ReaderT Environment m a }
   deriving newtype
     ( Applicative
     , Functor
     , Monad
+    , MonadCatch
     , MonadIO
-    , MonadReader (Environment (App m))
+    , MonadReader Environment
+    , MonadThrow
+    , MonadTrans
     )
 
-instance MonadTrans App where
-  lift = App . lift
+deriving instance MonadCatch m => MonadCatch (Colog.LoggerT Colog.Message m)
+deriving instance MonadThrow m => MonadThrow (Colog.LoggerT Colog.Message m)
```

Comment:
@epicallan We need these for integration with downstream.

Do you think they should exist, and thus be pushed upstream? 

Feel also free to in the companion PR on our project verify that these are actually required and not a side effect of a mistake I did over there.

Reply:
Yes I think these should exist because we use `WithEnv m env` in case where `m` requires these instances within `strainge`. 

> Do you think they should exist, and thus be pushed upstream?

Can you elaborate by what you mean pusing upstream

Reply:
@epicallan I meant that these instances as they are oprphaned, probably better are defined in `co-log` which owns `LoggerT`.

Reply:
>  I meant that these instances as they are oprphaned, probably better are defined in co-log which owns LoggerT.

Yes, I think its a good move